src/psyclone/psyir/transformations/omp_task_trans.py

```diff
@@ -0,0 +1,158 @@
+#   and/or other materials provided with the distribution.
+#
+# * Neither the name of the copyright holder nor the names of its
+#   contributors may be used to endorse or promote products derived from
+#   this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+# COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Author A. B. G. Chalk STFC Daresbury Lab
+''' This module provides the OMPTaskTrans transformation.'''
+from psyclone.errors import GenerationError
+from psyclone.domain.common.transformations import KernelModuleInlineTrans
+from psyclone.psyGen import Kern
+from psyclone.psyir.transformations.fold_conditional_return_expressions_trans \
+        import FoldConditionalReturnExpressionsTrans
+from psyclone.psyir.transformations.parallel_loop_trans import\
+    ParallelLoopTrans
+from psyclone.psyir.nodes import CodeBlock, Call
+from psyclone.psyir.nodes import DynamicOMPTaskDirective
+from psyclone.psyir.transformations.inline_trans import InlineTrans
+from psyclone.psyir.transformations.transformation_error import \
+        TransformationError
+
+
+class OMPTaskTrans(ParallelLoopTrans):
+    ''' Apply an OpenMP Task Transformation to a Loop. The Loop must
+    be within an OpenMP Serial region (Single or Master) at codegen time.
+    '''
+
+    def __str__(self):
+        return "Adds an 'OMP TASK' directive to a statement"
+
+    @property
+    def name(self):
+        '''
+        :returns: the name of this transformation.
+        :rtype: str
+        '''
+        return "OMPTaskTrans"
+
+    def validate(self, node, options=None):
+        '''
+        Validity checks for input arguments.
+
+        :param node: the Loop node to validate.
+        :type node: :py:class:`psyclone.psyir.nodes.Loop`
+        :param options: a dictionary with options for transformations.
+        :type options: dict of string:values or None
+        '''
+        # Disallow CodeBlocks inside the region
+        if len(node.walk(CodeBlock)) > 0:
+            raise GenerationError(
+                "OMPTaskTransformation cannot be applied to a region "
+                "containing a code block")
+        super().validate(node, options)
```

Comment:
This method needs to call the various `validate` methods of the transformations used within `_inline_kernels` as we have a rule that if `validate` succeeds then it must be safe to call `apply`.

Reply:
@arporter I'm not sure how to do this. I can create an instance of each of the transformations and call `.validate` on each `kern`, but that won't work. I could go through the process of copying all of the entities which would need the operation performing and then validating them, but that seems like a lot of extra computation.

Is there an alternative that you can see? Otherwise I'll have to go with that for now.

Reply:
@arporter The copy option is fine I think, however I cannot get `loop_trans_test.py::test_all_loop_trans_base_validate` to pass with this implementation.

My current implementation is:
```
        # Disallow CodeBlocks inside the region
        if any(node.walk(CodeBlock)):
            raise GenerationError(
                "OMPTaskTransformation cannot be applied to a region "
                "containing a code block")

        # Check we can apply all the required transformations on any sub
        # nodes
        root_ancestor = node.root
        path_to_node = node.path_from(root_ancestor)
        routine_copy = root_ancestor.copy()
        node_copy = routine_copy
        for index in path_to_node:
            node_copy = node_copy.children[index]

        kerns = node_copy.walk(Kern)
        kintrans = KernelModuleInlineTrans()
        cond_trans = FoldConditionalReturnExpressionsTrans()
        intrans = InlineTrans()
        for kern in kerns:
            kintrans.validate(kern)
            cond_trans.validate(kern.get_kernel_schedule())
            kintrans.apply(kern)
            cond_trans.apply(kern.get_kernel_schedule())
            kern.lower_to_language_level()

        calls = node_copy.walk(Call)
        for call in calls:
            intrans.validate(call)
        super().validate(node, options)
```

however this test is currently failing with `PSyclone internal error: Routine Symbol 'bc_ssh_code' is not local, unresolved or imported.` in `inline_trans::validate`. I previously tried without applying `kintrans` and `cond_trans` but I figured without those being applied, inline trans could fail for more reasons (and indeed, i think without applying kintrans bc_ssh_code wasn't even known by PSyclone).

I can make that test pass by moving super().validate to before the in-node validation which makes that test pass, do you think that what I'm doing is otherwise reasonable?

Reply:
So is `super().validate()` resulting in `inline_trans.validate()` being called a second time (i.e. after you have inlined the kernel)?